samples/dnn/DaSiamRPN_visual_tracker.py

```diff
@@ -0,0 +1,330 @@
+    x_crop = get_subwindow_tracking(im, target_pos, p.instance_size, round(s_x), avg_chans)
+    x_crop = np.transpose(x_crop, (2, 0, 1))
+    x_crop = np.reshape(x_crop, (1, 3, 271, 271)).astype(np.float32)
+
+    target_pos, target_sz, score = tracker_eval(net, x_crop, target_pos, target_sz * scale_z, window, scale_z, p)
+
+    target_pos[0] = max(0, min(state['im_w'], target_pos[0]))
+    target_pos[1] = max(0, min(state['im_h'], target_pos[1]))
+    target_sz[0] = max(10, min(state['im_w'], target_sz[0]))
+    target_sz[1] = max(10, min(state['im_h'], target_sz[1]))
+
+    state['target_pos'] = target_pos
+    state['target_sz'] = target_sz
+    state['score'] = score
+
+    return state
+
+#posistion of center of the rectangle and it's size
+def cxy_wh_2_rect(pos, sz):
+    return np.array([pos[0]-sz[0]/2, pos[1]-sz[1]/2, sz[0], sz[1]])
+
+#evaluations with cropped images
+def tracker_eval(net, x_crop, target_pos, target_sz, window, scale_z, p):
+    net.setInput(x_crop)
+    outNames = net.getUnconnectedOutLayersNames()
+    outNames = ['66', '68']
+    delta, score = net.forward(outNames)
+
+    delta = np.transpose(delta, (1, 2, 3, 0))
+    delta = np.ascontiguousarray(delta, dtype = np.float32)
+    delta = np.reshape(delta, (4, -1))
+
+    score = np.transpose(score, (1, 2, 3, 0))
+    score = np.ascontiguousarray(score, dtype = np.float32)
+    score = np.reshape(score, (2, -1))
+    score = softmax(score)[1, :]
+
+    delta[0, :] = delta[0, :] * p.anchor[:, 2] + p.anchor[:, 0]
+    delta[1, :] = delta[1, :] * p.anchor[:, 3] + p.anchor[:, 1]
+    delta[2, :] = np.exp(delta[2, :]) * p.anchor[:, 2]
+    delta[3, :] = np.exp(delta[3, :]) * p.anchor[:, 3]
+
+    def change(r):
+        return np.maximum(r, 1./r)
+
+    def sz(w, h):
+        pad = (w + h) * 0.5
+        sz2 = (w + pad) * (h + pad)
+        return np.sqrt(sz2)
+
+    def sz_wh(wh):
+        pad = (wh[0] + wh[1]) * 0.5
+        sz2 = (wh[0] + pad) * (wh[1] + pad)
+        return np.sqrt(sz2)
+
+    s_c = change(sz(delta[2, :], delta[3, :]) / (sz_wh(target_sz)))
+    r_c = change((target_sz[0] / target_sz[1]) / (delta[2, :] / delta[3, :]))
+
+    penalty = np.exp(-(r_c * s_c - 1.) * p.penalty_k)
+    pscore = penalty * score
+
+    pscore = pscore * (1 - p.window_influence) + window * p.window_influence
+    best_pscore_id = np.argmax(pscore)
+
+    target = delta[:, best_pscore_id] / scale_z
+    target_sz = target_sz / scale_z
+    lr = penalty[best_pscore_id] * score[best_pscore_id] * p.lr
+
+    res_x = target[0] + target_pos[0]
+    res_y = target[1] + target_pos[1]
+
+    res_w = target_sz[0] * (1 - lr) + target[2] * lr
+    res_h = target_sz[1] * (1 - lr) + target[3] * lr
+
+    target_pos = np.array([res_x, res_y])
+    target_sz = np.array([res_w, res_h])
+
+    return target_pos, target_sz, score[best_pscore_id]
+
+def softmax(x):
+    y = np.copy(x)
+
+    for i in range(1805):
+        e_x = np.exp(x[: , i] - np.max(x[ : , i]))
+        y[ : , i] = e_x /e_x.sum()
```

Comment:
Try to make it with numpy - compute max values along all the columns.

Reply:
Sorry, I don't understand what exactly do you mean - are you talking about line 275?
e_x = np.exp(x[: , i] - **np.max(x[ : , i]**))